// router_js/dist/modules/utils.js
export const slice = Array.prototype.slice;
const hasOwnProperty = Object.prototype.hasOwnProperty;
/**
  Determines if an object is Promise by checking if it is "thenable".
**/
export function isPromise(p) {
    return p !== null && typeof p === 'object' && typeof p.then === 'function';
}
export function merge(hash, other) {
    for (let prop in other) {
        if (hasOwnProperty.call(other, prop)) {
            hash[prop] = other[prop];
        }
    }
}
/**
  @private

  Extracts query params from the end of an array
**/
export function extractQueryParams(array) {
    let len = array && array.length, head, queryParams;
    if (len && len > 0) {
        let obj = array[len - 1];
        if (isQueryParams(obj)) {
            queryParams = obj.queryParams;
            head = slice.call(array, 0, len - 1);
            return [head, queryParams];
        }
    }
    return [array, null];
}
function isQueryParams(obj) {
    return obj && hasOwnProperty.call(obj, 'queryParams');
}
/**
  @private

  Coerces query param properties and array elements into strings.
**/
export function coerceQueryParamsToString(queryParams) {
    for (let key in queryParams) {
        let val = queryParams[key];
        if (typeof val === 'number') {
            queryParams[key] = '' + val;
        }
        else if (Array.isArray(val)) {
            for (let i = 0, l = val.length; i < l; i++) {
                val[i] = '' + val[i];
            }
        }
    }
}
/**
  @private
 */
export function log(router, ...args) {
    if (!router.log) {
        return;
    }
    if (arguments.length === 2) {
        let [sequence, msg] = args;
        router.log('Transition #' + sequence + ': ' + msg);
    }
    else {
        let [msg] = args;
        router.log(msg);
    }
}
export function isParam(object) {
    return (typeof object === 'string' ||
        object instanceof String ||
        typeof object === 'number' ||
        object instanceof Number);
}
export function forEach(array, callback) {
    for (let i = 0, l = array.length; i < l && callback(array[i]) !== false; i++) {
        // empty intentionally
    }
}
export function trigger(router, handlerInfos, ignoreFailure, name, ...args) {
    if (router.triggerEvent) {
        router.triggerEvent(handlerInfos, ignoreFailure, [name, ...args]);
        return;
    }
    if (!handlerInfos) {
        if (ignoreFailure) {
            return;
        }
        throw new Error("Could not trigger event '" + name + "'. There are no active handlers");
    }
    let eventWasHandled = false;
    for (let i = handlerInfos.length - 1; i >= 0; i--) {
        let currentHandlerInfo = handlerInfos[i], currentHandler = currentHandlerInfo.handler;
        // If there is no handler, it means the handler hasn't resolved yet which
        // means that we should trigger the event later when the handler is available
        if (!currentHandler) {
            currentHandlerInfo.handlerPromise.then(function (resolvedHandler) {
                resolvedHandler.events[name].apply(resolvedHandler, args);
            });
            continue;
        }
        if (currentHandler.events && currentHandler.events[name]) {
            if (currentHandler.events[name].apply(currentHandler, args) === true) {
                eventWasHandled = true;
            }
            else {
                return;
            }
        }
    }
    // In the case that we got an UnrecognizedURLError as an event with no handler,
    // let it bubble up
    if (name === 'error' && args[0].name === 'UnrecognizedURLError') {
        throw args[0];
    }
    else if (!eventWasHandled && !ignoreFailure) {
        throw new Error("Nothing handled the event '" + name + "'.");
    }
}
export function getChangelist(oldObject, newObject) {
    let key;
    let results = {
        all: {},
        changed: {},
        removed: {},
    };
    merge(results.all, newObject);
    let didChange = false;
    coerceQueryParamsToString(oldObject);
    coerceQueryParamsToString(newObject);
    // Calculate removals
    for (key in oldObject) {
        if (hasOwnProperty.call(oldObject, key)) {
            if (!hasOwnProperty.call(newObject, key)) {
                didChange = true;
                results.removed[key] = oldObject[key];
            }
        }
    }
    // Calculate changes
    for (key in newObject) {
        if (hasOwnProperty.call(newObject, key)) {
            let oldElement = oldObject[key];
            let newElement = newObject[key];
            if (isArray(oldElement) && isArray(newElement)) {
                if (oldElement.length !== newElement.length) {
                    results.changed[key] = newObject[key];
                    didChange = true;
                }
                else {
                    for (let i = 0, l = oldElement.length; i < l; i++) {
                        if (oldElement[i] !== newElement[i]) {
                            results.changed[key] = newObject[key];
                            didChange = true;
                        }
                    }
                }
            }
            else if (oldObject[key] !== newObject[key]) {
                results.changed[key] = newObject[key];
                didChange = true;
            }
        }
    }
    return didChange ? results : undefined;
}
function isArray(obj) {
    return Array.isArray(obj);
}
export function promiseLabel(label) {
    return 'Router: ' + label;
}
// export function callHook(obj: any, _hookName: string, arg1?: unknown, arg2?: unknown) {
//   let hookName = resolveHook(obj, _hookName);
//   return hookName && obj[hookName].call(obj, arg1, arg2);
// }
